// @wikifactory/slate-trailing-block/dist/index.js
'use strict';

var Slate = require('slate');
var focusAtEnd = require('./focusAtEnd');

/**
 * Slate plugin to ensure a trailing block.
 * @param {Object} [opts] Options for the plugin
 * @param {Function} [opts.match] Match last block
 * @param {String} [opts.type='paragraph'] The type of the trailing block to insert
 * @return {Object}
 */

function TrailingBlock(opts) {
    opts = opts || {};
    opts.type = opts.type || 'paragraph';
    opts.match = opts.match || function (node) {
        return node.type === opts.type;
    };

    return {
        onChange: function onChange(editor, next) {
            var document = editor.value.document;

            var lastNode = document.nodes.last();
            if (lastNode && opts.match(lastNode)) {
                return next();
            }

            var lastIndex = document.nodes.size;
            var block = Slate.Block.create({
                type: opts.type,
                nodes: [Slate.Text.create()]
            });
            editor.insertNodeByKey(document.key, lastIndex, block);
        },

        changes: {
            focusAtEnd: focusAtEnd
        }
    };
}

module.exports = TrailingBlock;
