// @blink-mind/plugin-theme-selector/lib/main.es.js
import { ModelModifier, defaultTheme } from '@blink-mind/core';

var theme1 = {
    name: 'theme1',
    randomColor: false,
    background: 'rgb(250,245,205)',
    highlightColor: '#50C9CE',
    marginH: 50,
    marginV: 30,
    fontFamily: '',
    bold: false,
    italic: false,
    textAlign: 'left',
    rootTopic: {
        background: 'rgb(221, 170, 68)',
        color: '#fff',
        fontSize: '34px',
        borderRadius: '5px',
        padding: '16px 18px 16px 18px'
    },
    primaryTopic: {
        background: '#e8eaec',
        borderWidth: '1px',
        borderStyle: 'solid',
        borderColor: 'rgb(221, 170, 68)',
        borderRadius: '5px',
        color: '#ffffff',
        fontSize: '17px',
        padding: '10px 15px 10px 15px',
        linkStyle: {
            lineType: 'curve',
            lineWidth: 2,
            lineColor: 'rgb(187, 136, 34)'
        }
    },
    normalTopic: {
        background: '#fff',
        border: '1px solid #e8eaec',
        borderRadius: '20px',
        color: 'rgb(187, 136, 34)',
        fontSize: '13px',
        padding: '4px 10px',
        linkStyle: {
            lineType: 'round',
            lineRadius: 5,
            lineWidth: 1,
            lineColor: 'rgb(187, 136, 34)'
        }
    }
};

var theme2 = {
    name: 'theme2',
    randomColor: false,
    background: '#f8f8f8',
    highlightColor: '#50C9CE',
    marginH: 45,
    marginV: 10,
    fontFamily: '',
    bold: false,
    italic: false,
    textAlign: 'left',
    rootTopic: {
        background: '#50C18A',
        color: '#fff',
        fontSize: '34px',
        borderRadius: '5px',
        padding: '16px 18px 16px 18px',
        linkStyle: {
            lineType: 'curve',
            lineWidth: 2,
            lineColor: 'rgb(113, 203, 45)'
        }
    },
    primaryTopic: {
        background: '#ffffff',
        borderWidth: '1px',
        borderStyle: 'solid',
        borderColor: 'rgb(221, 170, 68)',
        borderRadius: '5px',
        color: 'rgb(103,103,103)',
        fontSize: '14px',
        padding: '6px 10px 5px 10px',
        linkStyle: {
            lineType: 'curve',
            lineWidth: 2,
            lineColor: '#43a9ff'
        }
    },
    normalTopic: {
        background: '#fff',
        borderRadius: '5px',
        color: 'rgb(103,103,103)',
        fontSize: '13px',
        padding: '3px 9px 4px',
        boxShadow: '1px 1px 1px #ccc',
        linkStyle: {
            lineType: 'round',
            lineRadius: 5,
            lineWidth: 1,
            lineColor: '#43a9ff'
        }
    }
};

var theme3 = {
    name: 'theme3',
    randomColor: false,
    background: '#A9DEF9',
    highlightColor: '#50C9CE',
    marginH: 50,
    marginV: 20,
    fontFamily: '',
    bold: false,
    italic: false,
    textAlign: 'left',
    rootTopic: {
        background: '#FF99C8',
        color: '#fff',
        fontSize: '34px',
        borderRadius: '5px',
        padding: '16px 18px 16px 18px',
        linkStyle: {
            lineType: 'curve',
            lineWidth: 2,
            lineColor: 'rgb(113, 203, 45)'
        }
    },
    primaryTopic: {
        background: '#FCF6BD',
        borderWidth: '1px',
        borderStyle: 'solid',
        borderColor: '#E4C1F9',
        borderRadius: '5px',
        color: 'rgb(103,103,103)',
        fontSize: '14px',
        padding: '6px 10px 5px 10px',
        linkStyle: {
            lineType: 'curve',
            lineWidth: 2,
            lineColor: '#43a9ff'
        }
    },
    normalTopic: {
        background: '#E4FDE1',
        borderRadius: '5px',
        color: 'rgb(103,103,103)',
        fontSize: '13px',
        padding: '3px 9px 4px',
        boxShadow: '1px 1px 1px #ccc',
        linkStyle: {
            lineType: 'round',
            lineRadius: 5,
            lineWidth: 1,
            lineColor: '#43a9ff'
        }
    }
};

var theme4 = {
    name: 'theme3',
    randomColor: false,
    background: '#CCFBFE',
    highlightColor: '#50C9CE',
    marginH: 50,
    marginV: 20,
    fontFamily: '',
    bold: false,
    italic: false,
    textAlign: 'left',
    rootTopic: {
        background: '#CD8987',
        color: '#fff',
        fontSize: '34px',
        borderRadius: '5px',
        padding: '16px 18px 16px 18px',
        linkStyle: {
            lineType: 'curve',
            lineWidth: 2,
            lineColor: 'rgb(113, 203, 45)'
        }
    },
    primaryTopic: {
        background: '#CDACA1',
        borderWidth: '1px',
        borderStyle: 'solid',
        borderColor: '#E4C1F9',
        borderRadius: '5px',
        color: 'rgb(103,103,103)',
        fontSize: '14px',
        padding: '6px 10px 5px 10px',
        linkStyle: {
            lineType: 'line',
            lineWidth: 2,
            lineColor: '#43a9ff'
        }
    },
    normalTopic: {
        background: '#CDCACC',
        borderRadius: '5px',
        color: 'rgb(103,103,103)',
        fontSize: '13px',
        padding: '3px 9px 4px',
        boxShadow: '1px 1px 1px #ccc',
        linkStyle: {
            lineType: 'line',
            lineRadius: 5,
            lineWidth: 1,
            lineColor: '#43a9ff'
        }
    }
};

function ThemeSelectorPlugin() {
    var themeMap = new Map([
        ['default', defaultTheme],
        ['theme1', theme1],
        ['theme2', theme2],
        ['theme3', theme3],
        ['theme4', theme4]
    ]);
    return {
        getAllThemes: function (props) {
            return themeMap;
        },
        setTheme: function (props) {
            var model = props.model, controller = props.controller, themeKey = props.themeKey;
            var allTheme = controller.run('getAllThemes', props);
            if (!allTheme.has(themeKey)) {
                throw new Error('setTheme: the theme key is not correct!');
            }
            var theme = allTheme.get(themeKey);
            var newModel = ModelModifier.setTheme({ model: model, theme: theme });
            controller.change(newModel);
        }
    };
}

export { ThemeSelectorPlugin, theme1, theme2, theme3, theme4 };
